Allow set optional `pre_hook` command while defining pattern

vitalk opened the pull request:
This allow to execute user-defined command when pattern match found, as per #1. The function uses `<f-args>` to hold arguments list, so each tab and white space should be escaped with one backslash (see `:h <f-args>` for more details).

For example it’s possible to explicitly set some variables after filetype is set:

``` viml
AddShebangPattern  sh  ^#!/bin/sh  let\ g:is_sh=1
```

---

@oryband, can you please add any feedback about this feature?

vitalk committed:
Allow set optional `post_hook` command while defining pattern

This allow to execute user-defined command when pattern match found,
as per #1. The function uses <f-args> to hold arguments list, so each tab and
white space should be escaped with one backslash (see :h <f-args> for more
details).

For example it’s possible to explicitly set some variables after filetype is
set:

    AddShebangPattern  sh  ^#!/bin/sh  let\ g:is_sh=1

oryband commented:
This needs to be tested with switching between multiple bash and sh buffers, to see if the syntax highlighting changes etc.

vitalk committed:
Execute hook before set filetype to ensure syntax is changed

vitalk committed:
Add extra shebang pattern to detect sh/bash

Ensure the b:is_sh and b:is_bash variables set/unset according with shebang
pattern. Closes #1.
